fivem_script/tokovoip_script/src/c_main.lua

```diff
@@ -18,9 +18,136 @@ local targetPed;
 local useLocalPed = true;
 local isRunning = false;
 local scriptVersion = "1.3.2";
+local animStates = {}
+local displayingPluginScreen = false;
+local HeadBone = 0x796e;
+
+--------------------------------------------------------------------------------
+--	Plugin functions
+--------------------------------------------------------------------------------
+
+-- Handles the talking state of other players to apply talking animation to them
+local function setPlayerTalkingState(player, playerServerId)
+	local talking = tonumber(getPlayerData(playerServerId, "voip:talking"));
+	if (animStates[playerServerId] == 0 and talking == 1) then
+		PlayFacialAnim(GetPlayerPed(player), "mic_chatter", "mp_facial");
+	elseif (animStates[playerServerId] == 1 and talking == 0) then
+		PlayFacialAnim(GetPlayerPed(player), "mood_normal_1", "facials@gen_male@base");
+	end
+	animStates[playerServerId] = talking;
+end
+
+RegisterNUICallback("updatePluginData", function(data)
+	local payload = data.payload;
+	if (voip[payload.key] == payload.data) then return end
+	voip[payload.key] = payload.data;
+	setPlayerData(voip.serverId, "voip:" .. payload.key, voip[payload.key], true);
+	voip:updateConfig();
+	voip:updateTokoVoipInfo(true);
+end);
+
+-- Receives data from the TS plugin on microphone toggle
+RegisterNUICallback("setPlayerTalking", function(data)
+	voip.talking = tonumber(data.state);
+
+	if (voip.talking == 1) then
+		setPlayerData(voip.serverId, "voip:talking", 1, true);
+		PlayFacialAnim(GetPlayerPed(PlayerId()), "mic_chatter", "mp_facial");
+	else
+		setPlayerData(voip.serverId, "voip:talking", 0, true);
+		PlayFacialAnim(PlayerPedId(), "mood_normal_1", "facials@gen_male@base");
+	end
+end)
+
+local function clientProcessing()
+	local playerList = voip.playerList;
+	local usersdata = {};
+	local localHeading;
+	local ped = PlayerPedId()
+
+	if (voip.headingType == 1) then
+		localHeading = math.rad(GetEntityHeading(ped));
+	else
+		localHeading = math.rad(GetGameplayCamRot().z % 360);
+	end
+	local localPos;
+
+	if useLocalPed then
+		localPos = GetPedBoneCoords(ped, HeadBone);
+	else
+		localPos = GetPedBoneCoords(targetPed, HeadBone);
+	end
+
+	for _, player in pairs(playerList) do
```

Comment:
You should be able to use a standard for loop here. GetActivePlayers() seems to return a contigous array of player ids and pairs is a lot slower. In a quick test on a local server it seems to work and saves some cpu time

Something like
```lua
for i=1, #playerList do
  local player = playerList[i];
  local playerServerId = GetPlayerServerId(player);
  ...
```